flink-ml-core/src/main/java/org/apache/flink/ml/util/QuantileSummary.java

```diff
@@ -0,0 +1,359 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.util;
+
+import org.apache.flink.util.Preconditions;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Helper class to compute an approximate quantile summary. This implementation is based on the
+ * algorithm proposed in the paper: "Space-efficient Online Computation of Quantile Summaries" by
+ * Greenwald, Michael and Khanna, Sanjeev. (https://doi.org/10.1145/375663.375670)
+ */
+public class QuantileSummary {
+
+    private static final int BUFFER_SIZE = 10000;
```

Comment:
How is the default value for `BUFFER_SIZE` chosen? It seems that the corresponding value in Spark is 5000. [1]

By the way, `defaultHeadSize` seems a better name for this.

[1] https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala#L338

Reply:
I agree that `defaultHeadSize` is better. As for the default value, I think `10000` is enough instead of `50000` of Spark, because we don't want to allocate so much memory for each QuantileSummary.

Reply:
Hmm.. Is there any technical reasons for choosing `10000` instead of `50000`? If we want to avoid OOM, then why don't you choose `1000`?

Reply:
Of cause `50000` is also acceptable, in that way each QuantileSummary needs `50000 * 8B + 10000 * 24B = 640k` memory, a TM with 2GB mem can process 3000 features, if exceeded, the parallelism needs to be expanded.